src/o365/spo/commands/tenant/tenant-recyclebinitem-restore.ts

```diff
@@ -0,0 +1,149 @@
+import request from '../../../../request';
+import commands from '../../commands';
+import SpoCommand from '../../../base/SpoCommand';
+import GlobalOptions from '../../../../GlobalOptions';
+import { CommandOption, CommandCancel } from '../../../../Command';
+
+const vorpal: Vorpal = require('../../../../vorpal-init');
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  url: string;
+  wait: boolean;
+}
+
+interface RestSpoOperation {
+  HasTimedout: boolean;
+  IsComplete: boolean;
+  PollingInterval: number;
+}
+
+class SpoTenantRecycleBinItemRestoreCommand extends SpoCommand {
+  private timeout?: NodeJS.Timer;
+
+  public get name(): string {
+    return commands.TENANT_RECYCLEBINITEM_RESTORE;
+  }
+
+  public get description(): string {
+    return 'Restores the specified deleted Site Collection from Site Collection Recycle Bin';
+  }
+
+  public commandAction(cmd: CommandInstance, args: CommandArgs, cb: (err?: any) => void): void {
+    const wait: boolean = args.options.wait || false;
+    let spoAdminUrl: string;
+
+    this.getSpoAdminUrl(cmd, this.debug)
+    .then((adminUrl: string): Promise<RestSpoOperation> => {
+      spoAdminUrl = adminUrl;
+      return this.restoreDeletedSite(args.options.url, spoAdminUrl, cmd);
+    })
+    .then((response: RestSpoOperation): void => {
+      if (!response.HasTimedout && response.IsComplete) {
+        if (this.verbose) {
+          cmd.log('Site Collection restored');
+        }
+      }
+      else if (wait) {
+        this.waitForRestoreDeletedSite(args.options.url, spoAdminUrl, cmd, response);
+      }
+      else {
+        cmd.log('Site Collection has not been restored');
+      }
+
+      cb();
+    }, (err: any): void => this.handleRejectedODataJsonPromise(err, cmd, cb));
+  }
+
+  public cancel(): CommandCancel {
+    return (): void => {
+      if (this.timeout) {
+        clearTimeout(this.timeout);
+      }
+    }
+  }
+
+  private waitForRestoreDeletedSite(siteToRestoreUrl: string, spoAdminUrl: string, cmd: CommandInstance, response: RestSpoOperation): void {
+    this.timeout = setTimeout(() => {
+      this.restoreDeletedSite(siteToRestoreUrl, spoAdminUrl, cmd)
+      .then((respRetry: RestSpoOperation) => {
+        this.waitForRestoreDeletedSite(siteToRestoreUrl, spoAdminUrl, cmd, respRetry);
+      })
+      .catch((err) => {
+        cmd.log('Site Collection has not been restored');
+      });
+    }, response.PollingInterval);
+  }
+
+  private restoreDeletedSite(siteToRestoreUrl: string, spoAdminUrl: string, cmd: CommandInstance): Promise<RestSpoOperation> {
+    if (this.verbose) {
+      cmd.log('Restoring the deleted Site Collection');
+    }
+
+    if (this.debug) {
+      cmd.log(`Submitting POST request to ${spoAdminUrl}/_api/Microsoft.Online.SharePoint.TenantAdministration.Tenant/RestoreDeletedSite`);
```

Comment:
Including the URL in the debug statement is not necessary, because when you use debug mode, the underlying request handler already outputs the full request information.

Reply:
Right. So better log it in verbose mode or just remove the line ?

Reply:
Better remove it here to avoid duplication